File: Java/8 - Binary Trees and Binary Search Trees - Greg Hogg DSA Course Materials Lecture 8.java
// Binary Trees and Binary Search Trees - Greg Hogg DSA Course Materials Lecture 8

// Binary Tree Functions
public class Main {
    public static void main(String[] args) {
        TreeNode A = new TreeNode(1);
        TreeNode B = new TreeNode(2);
        TreeNode C = new TreeNode(3);
        TreeNode D = new TreeNode(4);
        TreeNode E = new TreeNode(5);
        TreeNode F = new TreeNode(10);

        A.left = B;
        A.right = C;
        B.left = D;
        B.right = E;
        C.left = F;

        System.out.println("Root Node: " + A);

        // Recursive Pre Order Traversal (DFS)
        System.out.println("Pre Order Traversal:");
        preOrder(A);

        // Recursive In Order Traversal (DFS)
        System.out.println("In Order Traversal:");
        inOrder(A);

        // Iterative Pre Order Traversal (DFS)
        System.out.println("Iterative Pre Order Traversal:");
        preOrderIterative(A);

        // Level Order Traversal (BFS)
        System.out.println("Level Order Traversal:");
        levelOrder(A);

        // Search for a value in the tree
        int target = 11;
        boolean found = search(A, target);
        System.out.println("Value " + target + " found in tree: " + found);
    }

    static class TreeNode {
        int val;
        TreeNode left, right;

        TreeNode(int val) {
            this.val = val;
        }

        public String toString() {
            return Integer.toString(val);
        }
    }

    // Recursive Pre Order Traversal (DFS)
    public static void preOrder(TreeNode node) {
        if (node == null) return;
        System.out.println(node);
        preOrder(node.left);
        preOrder(node.right);
    }

    // Recursive In Order Traversal (DFS)
    public static void inOrder(TreeNode node) {
        if (node == null) return;
        inOrder(node.left);
        System.out.println(node);
        inOrder(node.right);
    }

    // Iterative Pre Order Traversal (DFS)
    public static void preOrderIterative(TreeNode node) {
        if (node == null) return;
        java.util.Stack<TreeNode> stack = new java.util.Stack<>();
        stack.push(node);

        while (!stack.isEmpty()) {
            TreeNode current = stack.pop();
            System.out.println(current);

            if (current.right != null) stack.push(current.right);
            if (current.left != null) stack.push(current.left);
        }
    }

    // Level Order Traversal (BFS)
    public static void levelOrder(TreeNode node) {
        if (node == null) return;
        java.util.Queue<TreeNode> queue = new java.util.LinkedList<>();
        queue.add(node);

        while (!queue.isEmpty()) {
            TreeNode current = queue.poll();
            System.out.println(current);

            if (current.left != null) queue.add(current.left);
            if (current.right != null) queue.add(current.right);
        }
    }

    // Search for a value in the tree (DFS)
    public static boolean search(TreeNode node, int target) {
        if (node == null) return false;
        if (node.val == target) return true;
        return search(node.left, target) || search(node.right, target);
    }
}



// Binary Search Trees:

public class Main {
    public static void main(String[] args) {
        TreeNode A2 = new TreeNode(5);
        TreeNode B2 = new TreeNode(1);
        TreeNode C2 = new TreeNode(8);
        TreeNode D2 = new TreeNode(-1);
        TreeNode E2 = new TreeNode(3);
        TreeNode F2 = new TreeNode(7);
        TreeNode G2 = new TreeNode(9);

        A2.left = B2;
        A2.right = C2;
        B2.left = D2;
        B2.right = E2;
        C2.left = F2;
        C2.right = G2;

        System.out.println("BST Root Node: " + A2);

        // In Order Traversal (DFS)
        System.out.println("BST In Order Traversal:");
        inOrder(A2);

        // Search in BST
        int target = -1;
        boolean found = searchBST(A2, target);
        System.out.println("Value " + target + " found in BST: " + found);
    }

    // Node class for the Binary Search Tree
    static class TreeNode {
        int val;
        TreeNode left, right;

        TreeNode(int val) {
            this.val = val;
        }

        public String toString() {
            return Integer.toString(val);
        }
    }

    // In Order Traversal (DFS)
    public static void inOrder(TreeNode node) {
        if (node == null) return;
        inOrder(node.left);
        System.out.println(node);
        inOrder(node.right);
    }

    // Search in BST - O(log n) time, O(log n) space
    public static boolean searchBST(TreeNode node, int target) {
        if (node == null) return false;
        if (node.val == target) return true;
        if (target < node.val) return searchBST(node.left, target);
        else return searchBST(node.right, target);
    }
}

File: Python/8 - Binary Trees and Binary Search Trees - Greg Hogg DSA Course Materials Lecture 8.py
# Binary Trees and Binary Search Trees - Greg Hogg DSA Course Materials Lecture 8

# The Colab Notebook can be viewed here: https://colab.research.google.com/drive/1d0z8pA4Swkk-_bKZkaKpuRQvCNYecxNA


# Binary Trees
class TreeNode:
  def __init__(self, val, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

  def __str__(self):
    return str(self.val)

#        1
#     2    3
#   4  5  10

A = TreeNode(1)
B = TreeNode(2)
C = TreeNode(3)
D = TreeNode(4)
E = TreeNode(5)
F = TreeNode(10)

A.left = B
A.right = C
B.left = D
B.right = E
C.left = F

print(A)

# Recursive Pre Order Traversal (DFS) Time: O(n), Space: O(n)
def pre_order(node):
  if not node:
    return

  print(node)
  pre_order(node.left)
  pre_order(node.right)

pre_order(A)

# Recursive In Order Traversal (DFS) Time: O(n), Space: O(n)
def in_order(node):
  if not node:
    return

  in_order(node.left)
  print(node)
  in_order(node.right)

in_order(A)

# Iterative Pre Order Traversal (DFS) Time: O(n), Space: O(n)
def pre_order_iterative(node):
  stk = [node]

  while stk:
    node = stk.pop()
    if node.right: stk.append(node.right)
    if node.left: stk.append(node.left)
    print(node)

pre_order_iterative(A)

# Level Order Traversal (BFS) Time: O(n), Space: O(n)
from collections import deque

def level_order(node):
  q = deque()
  q.append(node)

  while q:
    node = q.popleft()
    print(node)
    if node.left: q.append(node.left)
    if node.right: q.append(node.right)

level_order(A)

# Check if Value Exists (DFS) Time: O(n), Space: O(n)
def search(node, target):
  if not node:
    return False

  if node.val == target:
    return True

  return search(node.left, target) or search(node.right, target)

print(search(A, 11))

# Binary Search Trees (BSTs)

#       5
#    1    8
#  -1 3  7 9

A2 = TreeNode(5)
B2 = TreeNode(1)
C2 = TreeNode(8)
D2 = TreeNode(-1)
E2 = TreeNode(3)
F2 = TreeNode(7)
G2 = TreeNode(9)

A2.left, A2.right = B2, C2
B2.left, B2.right = D2, E2
C2.left, C2.right = F2, G2

print(A2)

in_order(A2)

# Time: O(log n), Space: O(log n)
def search_bst(node, target):
  if not node:
    return False

  if node.val == target:
    return True

  if target < node.val: return search_bst(node.left, target)
  else: return search_bst(node.right, target)

print(search_bst(A2, -1))
